servers/rendering/renderer_rd/forward_mobile/scene_shader_forward_mobile.cpp

```diff
@@ -141,13 +141,12 @@ void SceneShaderForwardMobile::ShaderData::set_code(const String &p_code) {
 
 	actions.uniforms = &uniforms;
 
-	SceneShaderForwardMobile *shader_singleton = (SceneShaderForwardMobile *)SceneShaderForwardMobile::singleton;
-
-	Error err = shader_singleton->compiler.compile(RS::SHADER_SPATIAL, code, &actions, path, gen_code);
+	MutexLock lock(SceneShaderForwardMobile::singleton_mutex);
```

Comment:
Can you remind we why we are locking a mutex here? I don't there is anything in `compile()` that we need to worry about

Reply:
Yes, compiler itself is not thread-safe. It uses internal members and getters. We'd need to instance a compiler specific to the thread to be able to use it.

Reply:
I've been able to actually make the compile method const with some refactoring. It seems like the fact it was was mostly a carry-over from some older code.

Reply:
I gave it a shot but making this const is a bit difficult. It requires a large amount of refactoring and even the inputs to the function can be modified, and some of those inputs are also members of the shader itself.

I can dedicate more time to it if necessary but leaving the mutex is the safe way to go.

Reply:
No need to refactor the compiler. Thanks for the explanation, that's all I needed